lib/plug/cowboy/telemetry.ex

```diff
@@ -0,0 +1,89 @@
+defmodule Plug.Cowboy.Telemetry do
+  def init(stream_id, req, opts) do
+    system_time = System.system_time()
+    start_time = System.monotonic_time()
+
+    :telemetry.execute([:plug_cowboy, :stream_handler, :start], %{system_time: system_time}, %{
+      host: req.host,
+      method: req.method,
+      request_path: req.path,
+      port: req.port,
+      remote_ip: req.peer |> elem(0),
+      scheme: req.scheme,
+      req_headers: req.headers |> Map.to_list(),
+      query_string: req.qs
+    })
+
+    {commands, next} = :cowboy_stream.init(stream_id, req, opts)
+    {commands, %{next: next, start_time: start_time}}
+  end
+
+  def info(stream_id, info, %{next: next, start_time: start_time}) do
+    end_time = System.monotonic_time()
+
+    case info do
+      {:response, <<status::binary-size(3)>> <> _message, headers, body} ->
+        :telemetry.execute(
+          [:plug_cowboy, :stream_handler, :stop],
+          %{duration: end_time - start_time},
+          %{
+            status: status |> String.to_integer(),
+            resp_headers: headers |> Map.to_list(),
+            body: body
+          }
+        )
+
+      {:EXIT, _pid, {{{reason, stacktrace}, _init_call}, _exit_stack}} ->
```

Comment:
This information is also available in the `terminate` callback wrapped in an `internal_error` reason tuple:

https://github.com/ninenines/cowboy/blob/de955079f90d52cc626baebb52154f116c1ab5dd/src/cowboy_stream.erl#L49-L54

```erlang
-type reason() :: normal | switch_protocol
	| {internal_error, timeout | {error | exit | throw, any()}, human_reason()}
	| {socket_error, closed | atom(), human_reason()}
	| {stream_error, cow_http2:error(), human_reason()}
	| {connection_error, cow_http2:error(), human_reason()}
	| {stop, cow_http2:frame() | {exit, any()}, human_reason()}.
```

I'm still trying to think through which situations of all these are ones we need to cover

Reply:
One possibility is that if we want this to be more of a `cowboy_telemetry_h`, we could map the events directly to cowboy terminology, and execute the callbacks with the raw data that are handed.

This would mean it'd be the responsibility of the handler to understand them and generate data (vs us translating this into Span shaped events)

I don't know if that's the ideal situation, but it is plausible...

Reply:
> This would mean it'd be the responsibility of the handler to understand them and generate data (vs us translating this into Span shaped events)

I agree. Maybe there is a concern about exposing too much of Cowboy's behaviour. But we can definitely strike a balance.

Reply:
Just to clarify, which direction would you suggest we take this? Generate telemetry style spans or forward cowboy details?

My take is that the internal cowboy details are somewhat complex and it's probably going to be difficult for folks to "get it right" if they have to translate them. The code now ignores most callback execution & parses out response & error structures, so I lean towards generating nice clean spans..

Reply:
I would go with spans, because we want to make them easy to consume, but we probably don't need to break down the metadata so much and forward Cowboy's stuff.

Reply:
In particular, I would go with `{:exit, reason}` for the exception event metadata. Could that work?

Reply:
I tried to follow the span "spec" https://github.com/beam-telemetry/telemetry/issues/57 for exception so did a map of kind, reason, stacktrace here

Reply:
I see. Those conversion rules are a bit more complicated. For example, you may not have a init_call, you can have false positives (you have to check if the stacktrace is actually a stacktrace, etc. My thinking is that, from the perspective of this process, they are all exits. The actual logic to handle is going to be around 100LOC and you can find it in the Elixir Exception module.

Reply:
Updated to send the whole `reason` without unpacking it...